[threaded-animation-resolution] webanimations/transform-accelerated-animation-finishes-before-removal.html is a failure

graouts opened the pull request:
#### c409399155163c223f7bb1bca5d54e71a8724b99
<pre>
[threaded-animation-resolution] webanimations/transform-accelerated-animation-finishes-before-removal.html is a failure
<a href="https://bugs.webkit.org/show_bug.cgi?id=269278">https://bugs.webkit.org/show_bug.cgi?id=269278</a>
<a href="https://rdar.apple.com/122856408">rdar://122856408</a>

Reviewed by Dean Jackson.

We need to ensure keyframe effects that are accelerated using threaded animation resolution cause the target element
to remain on a layer until the effect has completed. In the case of threaded animation resolution, because accelerated
effects run in the UIProcess, we cannot simply rely on whether the keyframe effect stack has a running effect indicating
it is accelerated, because it will no longer be &quot;running&quot; as soon as timing in the WebProcess resolves to such a state.

What we want is to ensure the target element remains on a layer until after that state has been propagated to the UIProcess.
To do so, we now set a list of all accelerated effects on the `KeyframeEffectStack` when we compute the latest accelerated
effect stack in RenderLayerBacking::updateAcceleratedEffectsAndBaseValues(). Then the keyframe effect stack can be queried
for the presence of at least one accelerated effect using `KeyframeEffectStack::hasAcceleratedEffects()`. We call this new
function in two spots:

    1. `RenderLayerCompositor::requiresCompositingForAnimation()`
    2. `Styleable::hasRunningAcceleratedAnimations()`

The latter was formerly called `Styleable::runningAnimationsAreAllAccelerated()` which, while named accurately, was not
doing what was required, which was to identify whether at least one animation was being run with acceleration. This method
is called under `RenderBox::requiresLayer()` and `RenderInline::requiresLayer()`.

This addresses the test regression with threaded animation resolution enabled.

* Source/WebCore/animation/KeyframeEffectStack.cpp:
(WebCore::KeyframeEffectStack::hasAcceleratedEffects const):
* Source/WebCore/animation/KeyframeEffectStack.h:
(WebCore::KeyframeEffectStack::setAcceleratedEffects):
* Source/WebCore/rendering/RenderBoxModelObject.cpp:
(WebCore::RenderBoxModelObject::hasRunningAcceleratedAnimations const):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateAcceleratedEffectsAndBaseValues):
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::requiresCompositingForAnimation const):
* Source/WebCore/style/Styleable.cpp:
(WebCore::Styleable::hasRunningAcceleratedAnimations const):
(WebCore::Styleable::runningAnimationsAreAllAccelerated const): Deleted.
* Source/WebCore/style/Styleable.h:

Canonical link: <a href="https://commits.webkit.org/275178@main">https://commits.webkit.org/275178@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/aeb283c8ededf43f0119ea402e2e405fc123c14e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40376 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19388 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42754 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42922 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/36465 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/42683 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22331 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16717 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/33488 "Found 1 new test failure: media/track/media-element-enqueue-event-crash.html (failure)") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40950 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16335 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34810 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14072 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/14129 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35775 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44199 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36597 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36098 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/39840 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/15191 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12438 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/38113 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16810 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9227 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/16859 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/16454 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/style/Styleable.cpp:
```diff
@@ -194,6 +195,11 @@ bool Styleable::isRunningAcceleratedTransformAnimation() const
 bool Styleable::runningAnimationsAreAllAccelerated() const
 {
+    if (auto* renderer = dynamicDowncast<RenderLayerModelObject>(this->renderer())) {
+        if (renderer->hasAcceleratedEffects())
```
We only use threaded animations when all animations on an element are accelerated? Where is this decision made?

smfr replied:
I'm also a little surprised that any change is needed here. Why aren't threaded animations represented in `keyframeEffectStack()` which would make the below code take care of this?

graouts replied:
> We only use threaded animations when all animations on an element are accelerated? Where is this decision made?

No, we don't. We use threaded animations as long as properties set on that element doesn't prevent us from doing so. Much of that code hasn't landed yet as I still need to refactor it a bit but some examples would be:

- animating a width-dependent `translate` value while also animating `width`
- animating `filter` with a discrete interval yielding mis-matching filter lists
- animating certain Motion Path properties which require computations we cannot perform in the UI Process

I don't recall why this method is named the way it is since the single call site for it `RenderBoxModelObject::hasRunningAcceleratedAnimations()` which certainly does not sound like it's exclusive. The only situation where we should be interested in whether **all** animations are running accelerated is when we want to avoid scheduling animations in the Web Process.

> I'm also a little surprised that any change is needed here. Why aren't threaded animations represented in `keyframeEffectStack()` which would make the below code take care of this?

The way the threaded animation resolution code works is that we create a separate `AcceleratedEffectStack` containing accelerated effects that will get uploaded to the UIProcess. The managed objects are different, `AcceleratedEffect` v. `KeyframeEffect`. They share some traits, for instance both support resolving timing and blending keyframes, but they are different since one is designed to animate `CALayer` properties and the other CSS properties. One is designed to have only the minimal number of properties to run in the UI Process, the other support all the complexity of the Web Animations API.

The code below is really the "old way" where calling `Keyframe::isRunningAccelerated()` would indicate whether a `CAAnimation` is running for that effect. The new code queries the accelerated effect stack to see if it exists and has effects, meaning there are animations running in the UI Process for it.

grorg commented on Source/WebCore/style/Styleable.cpp:
```diff
@@ -245,18 +245,11 @@ bool Styleable::isRunningAcceleratedTransformAnimation() const
 }
 
-bool Styleable::runningAnimationsAreAllAccelerated() const
+bool Styleable::hasRunningAcceleratedAnimations() const
```
Maybe `isRunningAnAcceleratedAnimation`?

grorg replied:
Or `hasAcceleratedAnimation` or `hasActiveAcceleratedAnimation`? Something about the `hasRunning` bit is confusing to me.

graouts replied:
The term [running](https://drafts.csswg.org/web-animations-1//#play-state-running) is important because the animation needs to be producing an interpolating value from frame to frame, so being [active](https://drafts.csswg.org/web-animations-1//#animation-effect-active-phase) is not sufficient. I’m not sure the suggestions are much of an improvement, so I’ll land as-is for now, but feel free to make more suggestions and this can be addressed in a follow-up.

graouts committed:
[threaded-animation-resolution] webanimations/transform-accelerated-animation-finishes-before-removal.html is a failure
https://bugs.webkit.org/show_bug.cgi?id=269278
rdar://122856408

Reviewed by Dean Jackson.

We need to ensure keyframe effects that are accelerated using threaded animation resolution cause the target element
to remain on a layer until the effect has completed. In the case of threaded animation resolution, because accelerated
effects run in the UIProcess, we cannot simply rely on whether the keyframe effect stack has a running effect indicating
it is accelerated, because it will no longer be "running" as soon as timing in the WebProcess resolves to such a state.

What we want is to ensure the target element remains on a layer until after that state has been propagated to the UIProcess.
To do so, we now set a list of all accelerated effects on the `KeyframeEffectStack` when we compute the latest accelerated
effect stack in RenderLayerBacking::updateAcceleratedEffectsAndBaseValues(). Then the keyframe effect stack can be queried
for the presence of at least one accelerated effect using `KeyframeEffectStack::hasAcceleratedEffects()`. We call this new
function in two spots:

    1. `RenderLayerCompositor::requiresCompositingForAnimation()`
    2. `Styleable::hasRunningAcceleratedAnimations()`

The latter was formerly called `Styleable::runningAnimationsAreAllAccelerated()` which, while named accurately, was not
doing what was required, which was to identify whether at least one animation was being run with acceleration. This method
is called under `RenderBox::requiresLayer()` and `RenderInline::requiresLayer()`.

This addresses the test regression with threaded animation resolution enabled.

* Source/WebCore/animation/KeyframeEffectStack.cpp:
(WebCore::KeyframeEffectStack::hasAcceleratedEffects const):
* Source/WebCore/animation/KeyframeEffectStack.h:
(WebCore::KeyframeEffectStack::setAcceleratedEffects):
* Source/WebCore/rendering/RenderBoxModelObject.cpp:
(WebCore::RenderBoxModelObject::hasRunningAcceleratedAnimations const):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateAcceleratedEffectsAndBaseValues):
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::requiresCompositingForAnimation const):
* Source/WebCore/style/Styleable.cpp:
(WebCore::Styleable::hasRunningAcceleratedAnimations const):
(WebCore::Styleable::runningAnimationsAreAllAccelerated const): Deleted.
* Source/WebCore/style/Styleable.h:

Canonical link: https://commits.webkit.org/275178@main
